x-pack/plugins/timelines/common/utils/field_formatters.ts

```diff
@@ -51,117 +57,141 @@ export const isRuleParametersFieldOrSubfield = (field: string, prependField?: st
+  values,
+  originalValue: values,
+  isObjectArray,
+});
+
+const processGeoField = (
+  field: string,
+  item: unknown[],
+  fieldCategory: string
+): TimelineEventsDetailsItem => {
+  const formattedLocation = formatGeoLocation(item);
+  return createFieldItem(fieldCategory, field, formattedLocation, true);
+};
+
+const processSimpleField = (
+  dotField: string,
+  strArr: string[],
+  isObjectArray: boolean,
+  fieldCategory: string
+): TimelineEventsDetailsItem => createFieldItem(fieldCategory, dotField, strArr, isObjectArray);
+
+const processNestedFields = (
+  item: unknown,
+  dotField: string,
+  fieldCategory: string,
+  prependDotField: boolean
+): TimelineEventsDetailsItem[] => {
+  if (Array.isArray(item)) {
+    return item.flatMap((curr) =>
+      getDataFromFieldsHits(curr as Fields, prependDotField ? dotField : undefined, fieldCategory)
+    );
+  }
+
+  return getDataFromFieldsHits(
+    item as Fields,
+    prependDotField ? dotField : undefined,
+    fieldCategory
+  );
+};
+
+type DisjointFieldNames = 'ecs.version' | 'event.action' | 'event.kind' | 'event.original';
+
+// Memoized field maps
+const fieldMaps: EcsFieldMap &
+  Omit<TechnicalRuleFieldMap, DisjointFieldNames> &
+  ExperimentalRuleFieldMap = {
+  ...technicalRuleFieldMap,
+  ...ecsFieldMap,
+  ...legacyExperimentalFieldMap,
+};
+
 export const getDataFromFieldsHits = (
   fields: Fields,
   prependField?: string,
   prependFieldCategory?: string
-): TimelineEventsDetailsItem[] =>
-  Object.keys(fields).reduce<TimelineEventsDetailsItem[]>((accumulator, field) => {
+): TimelineEventsDetailsItem[] => {
+  const resultMap = new Map<string, TimelineEventsDetailsItem>();
+  const fieldNames = Object.keys(fields);
+  for (let i = 0; i < fieldNames.length; i++) {
+    const field = fieldNames[i];
     const item: unknown[] = fields[field];
-    const fieldCategory =
-      prependFieldCategory != null ? prependFieldCategory : getFieldCategory(field);
+    const fieldCategory = prependFieldCategory ?? getFieldCategory(field);
+    const dotField = prependField ? `${prependField}.${field}` : field;
+
+    // Handle geo fields
     if (isGeoField(field)) {
-      return [
-        ...accumulator,
-        {
-          category: fieldCategory,
-          field,
-          values: formatGeoLocation(item),
-          originalValue: formatGeoLocation(item),
-          isObjectArray: true, // important for UI
-        },
-      ];
+      const geoItem = processGeoField(field, item, fieldCategory);
+      resultMap.set(field, geoItem);
+      // eslint-disable-next-line no-continue
+      continue;
     }
+
     const objArrStr = toObjectArrayOfStrings(item);
     const strArr = objArrStr.map(({ str }) => str);
     const isObjectArray = objArrStr.some((o) => o.isObjectArray);
-    const dotField = prependField ? `${prependField}.${field}` : field;
 
-    // return simple field value (non-ecs object, non-array)
-    if (
-      !isObjectArray ||
-      (Object.keys({
```

Comment:
Just realized this ran on every single field within a reduce 🤯 🤯 🤯 . Thanks for pulling this out 🙏🏾